File: app/src/test/java/com/kira/kdownloader/settings/SettingsMigrationTest.java
package com.kira.kdownloader.settings;

import com.kira.kdownloader.settings.store.InMemoryKeyValueStore;

import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;

public class SettingsMigrationTest {
    @Test public void freshStoreIsStampedWithTheCurrentVersion() {
        InMemoryKeyValueStore store = new InMemoryKeyValueStore();
        SettingsMigration.migrateIfNeeded(store);
        assertEquals(SettingsKeys.CURRENT_VERSION, store.getInt(SettingsKeys.VERSION, -1));
    }

    @Test public void alreadyCurrentStoreIsLeftUnchanged() {
        Map<String, String> initial = new HashMap<>();
        initial.put(SettingsKeys.VERSION, Integer.toString(SettingsKeys.CURRENT_VERSION));
        InMemoryKeyValueStore store = new InMemoryKeyValueStore(initial);
        SettingsMigration.migrateIfNeeded(store);
        assertEquals(SettingsKeys.CURRENT_VERSION, store.getInt(SettingsKeys.VERSION, -1));
    }

    @Test public void aNewerStoreIsNotDowngraded() {
        int newer = SettingsKeys.CURRENT_VERSION + 5;
        Map<String, String> initial = new HashMap<>();
        initial.put(SettingsKeys.VERSION, Integer.toString(newer));
        InMemoryKeyValueStore store = new InMemoryKeyValueStore(initial);
        SettingsMigration.migrateIfNeeded(store);
        assertEquals(newer, store.getInt(SettingsKeys.VERSION, -1));
    }
}

File: app/src/test/java/com/kira/kdownloader/settings/SettingsMigrationTest.kt
package com.kira.kdownloader.settings

import com.kira.kdownloader.settings.store.InMemoryKeyValueStore
import org.junit.Assert.assertEquals
import org.junit.Test

class SettingsMigrationTest {

    @Test
    fun `fresh store is stamped with the current version`() {
        val store = InMemoryKeyValueStore()
        SettingsMigration.migrateIfNeeded(store)
        assertEquals(SettingsKeys.CURRENT_VERSION, store.getInt(SettingsKeys.VERSION, -1))
    }

    @Test
    fun `already-current store is left unchanged`() {
        val store = InMemoryKeyValueStore(
            mapOf(SettingsKeys.VERSION to SettingsKeys.CURRENT_VERSION.toString()),
        )
        SettingsMigration.migrateIfNeeded(store)
        assertEquals(SettingsKeys.CURRENT_VERSION, store.getInt(SettingsKeys.VERSION, -1))
    }

    @Test
    fun `a newer store is not downgraded`() {
        val newer = SettingsKeys.CURRENT_VERSION + 5
        val store = InMemoryKeyValueStore(mapOf(SettingsKeys.VERSION to newer.toString()))
        SettingsMigration.migrateIfNeeded(store)
        assertEquals(newer, store.getInt(SettingsKeys.VERSION, -1))
    }
}
